base/util.go

```diff
@@ -47,25 +47,16 @@ const (
 	kMaxDeltaTtlDuration = 60 * 60 * 24 * 30 * time.Second
 )
 
-// NonCancellableContext is here to stroe a context that is not cancellable. Used to explicitly state when a change from
+// NonCancellableContext is here to store a context that is not cancellable. Used to explicitly state when a change from
 // a cancellable context to a context withoutr contex is required
 type NonCancellableContext struct {
 	Ctx context.Context
 }
 
 // NewNonCancelCtx creates a new background context struct for operations that require a fresh context
-func NewNonCancelCtx() NonCancellableContext {
+func NewNonCancelCtx(ctx context.Context) NonCancellableContext {
```

Comment:
I would say that for ISGR we do want to reset this information (but we already do by resetting correlation ID).

For the requests on admin API to change db config I think that having the context information is helpful. Any error from `_applyConfig` synchronous API would be helpful to know if it came from config polling or a request.

Situations that are ambiguous - `DatabaseInitManager`. I think conceptually having this be associated with a request might make sense, it is first in to do an initialization.

Inside `_applyConfig` the code that we don't want to have a context associated with a request are:

- NewImportFilterFunction
- newEventHandlers

asyncOnline already resets its context https://github.com/couchbase/sync_gateway/blob/296de2ec955c18cc766ac76c0acbb978d7c0c614/rest/server_context.go#L1002

Inside `NewDatabaseContext` anything that returns logging for an error we'd want to have the logging on the request but some things do require a context reset.

- `newSequenceAllocator` spawns a goroutine that uses a context
- `NewChannelCacheForContext` starts Background tasks
- `NewCfgSg` will pick up context but reset correlation ID
- `UpdateSyncFn` will use a context for the `JSRunner` 
- `NewSgReplicateMgr` sets up a context as member of `sgReplicateManager resetting the correlation ID.

I think as a possible compromise, resetting the ctx here https://github.com/couchbase/sync_gateway/blob/296de2ec955c18cc766ac76c0acbb978d7c0c614/rest/server_context.go#L814 and also in ActiveReplicator.

I think we have to assume that a context will be cancelled once the request is ended. It's going to be really hard to spot either case if it is "wrong" because it might not be the trivial actions that trigger logging, e.g. enabling ISGR or import filter functions and you have to spot the issue carefully. It's probably not a huge deal for database config operations to lose request context but there can always be the question of whether this action was triggered by a POST/PUT or config polling.

Reply:
I am really not following how a user-request to create a database needs to make it this deep into the actual instance of a database, long afer the original create request has finished? Maybe we need to sync up on this so I can understand.

Reply:
Just seen your follow-up commits to back out of this change. I'd still like to understand what that would've given us, but we can revisit another time if you want.